### PR TITLE
Use Nvidia operator

### PR DESCRIPTION
**Pull request by joedborg**
This branch works on nodes without any Nvidia drives installed.  We are waiting on https://github.com/NVIDIA/gpu-operator/issues/126 to allow this to work on nodes already provisioned (essential for most MicroK8s users).

The API remains the same:

```
sudo snap install microk8s --classic
sudo microk8s enable gpu
```

This should fix issues we've been seeing with users on odd combinations of OS and driver version and has also allowed us to strip a bunch of build steps, which is nice.

### CODE REVIEW

**Comment by plchampigny** [comment]
When do you expect this to be merged? I'm waiting precisely for this fix!

**Comment by joedborg** [comment]
> When do you expect this to be merged? I'm waiting precisely for this fix!

Hey @plchampigny, it’ll be merged at soon as the operator is compatible with systems that have existing drivers.  Feel free to try out the branched snap downloading it from here https://github.com/ubuntu/microk8s/suites/1867608846/artifacts/36367479

**Comment by plchampigny** [comment]
Hello @joedborg, it is still not working for me:
image

I removed the nvidia driver from my host as it would be conflicting with the operator. I didnt have time to investigate further.

**Comment by joedborg** [comment]
@plchampigny Interesting.  When you have 5 minutes to spare, could you please share the logs from the feature discovery pod?  And, just to confirm, the node you're trying this on has no Nvidia driver or Kernel module installed?

**Comment by plchampigny** [comment]
I cannot see the container logs, but here's the error from describe:

`Error: failed to create containerd task: OCI runtime create failed: container_linux.go:370: starting container process caused: process_linux.go:459: container init caused: Running hook #0:: error running hook: exit status 1, stdout: , stderr: nvidia-container-cli.real: mount error: mount operation failed: /var/snap/microk8s/common/run/containerd/io.containerd.runtime.v2.task/k8s.io/gpu-feature-discovery/rootfs/proc/driver/nvidia/params/version/registry: no such file or directory: unknown`

**Comment by plchampigny** [comment]
> And, just to confirm, the node you're trying this on has no Nvidia driver or Kernel module installed?

@joedborg Prior to install the snap, I had the driver 460 installed. After installing the gpu addon and seeing there was an error with the nvidia daemon, I did uninstall the nvidia driver (which solved the gpu daemon pod). The error you see on my 2 lasts post is the error I get after I removed the driver.

**Comment by plchampigny** [comment]
@joedborg Well eventually it did work: 
image

**Comment by joedborg** [comment]
> @joedborg Well eventually it did work:
> I restarted the pods a couple of times.

Glad it came up eventually.  I've only got AWS to test on (with the GPU nodes) at the moment.   I'd be interested to see if you find this with any other nodes (especially green field nodes).

File: microk8s-resources/actions/disable.gpu.sh [hunk 1]
```diff
@@ -4,21 +4,8 @@ set -e
...
-
-run_with_sudo preserve_env snapctl restart "${SNAP_NAME}.daemon-containerd"
-containerd_up=$(wait_for_service containerd)
-if [[ $containerd_up == fail ]]
-then
-  echo "Containerd did not start on time. Proceeding."
-fi
-# Allow for some seconds for containerd processes to start
-sleep 10
+"$SNAP/microk8s-helm3.wrapper" uninstall gpu-operator
```
Comment (ktsakalozos) [inline microk8s-resources/actions/disable.gpu.sh:9]: We need to enable helm3 before using it, right?

File: microk8s-resources/actions/enable.gpu.sh [hunk 2]
```diff
@@ -4,31 +4,27 @@ set -e
...
+"$SNAP/microk8s-enable.wrapper" dns
+"$SNAP/microk8s-enable.wrapper" helm3
+
+echo "Installing NVIDIA Operator"
+
+"$SNAP/microk8s-helm3.wrapper" repo add nvidia https://nvidia.github.io/gpu-operator
+"$SNAP/microk8s-helm3.wrapper" repo update
+"$SNAP/microk8s-helm3.wrapper" install gpu-operator nvidia/gpu-operator \
+  --set operator.defaultRuntime=containerd \
+  --set toolkit.version=1.4.1-ubuntu16.04 \
```
Comment (joedborg) [inline microk8s-resources/actions/enable.gpu.sh:24]: @ktsakalozos FYI, this will probably have to change when we change core.

**Review by ktsakalozos — APPROVED** [review]
LGTM +1

**Comment by ktsakalozos** [comment]
Thank you for this work @joedborg

**Comment by treo** [comment]
As a note, it appears that this breaks GPU support on non LTS releases. 

> [remote_image.go:87] "Get ImageStatus from image service failed" err="rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial unix /var/snap/microk8s/common/run/containerd.sock: connect: no such file or directory\"" image="nvcr.io/nvidia/driver:460.32.03-ubuntu20.10"

**Pull request merged**